### PR TITLE
Sourcery methods

### PR DESCRIPTION
**Pull request by joesus**
## Checklist

- [x] I've read the [Contributing Guidelines](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I've completed the [Contributor License Agreement](https://developers.facebook.com/opensource/cla)
- [x] I've ensured that all existing tests pass and added tests (when/where necessary)
- [x] I've ensured that my code lints properly: (`swiftlint && swiftlint autocorrect --format`)
- [x] I've updated the documentation (when/where necessary) and [Changelog](CHANGELOG.md) (when/where necessary)
- [x] I've added the proper label to this pull request (e.g. `bug` for bug fixes)

## Pull Request Details

This is a more involved implementation of using Sourcery to expose Swift code to ObjcC.

## Test Plan

Should be able to build the added project ObjCCatalog without issue.

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1 @@
+./Codegen/Generated/* linguist-generated=true
```
Comment (codytwinton) [inline .gitattributes]: I think the `./` might be messing this up.

```suggestion
Codegen/Generated/* linguist-generated=true
```

File: .sourcery.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+sources:
+  - ./Sources/Core
+templates:
+  include:
+    - ./Codegen/Templates
+  exclude:
+    - ./Codegen/Templates/Shared
+output:
+  path: ./Codegen/Generated
```
Comment (codytwinton) [inline .sourcery.yml]: Code directory specificity?

```suggestion
  path: ./Codegen/Generated/FacebookCore
```

File: FacebookSwift.xcodeproj/xcshareddata/xcschemes/FacebookCore.xcscheme [hunk 3]
```diff
@@ -27,8 +27,36 @@
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
       shouldUseLaunchSchemeArgsEnv = "YES">
+      <PreActions>
+         <ExecutionAction
+            ActionType = "Xcode.IDEStandardExecutionActionsCore.ExecutionActionType.ShellScriptAction">
+            <ActionContent
+               title = "Run Script"
+               scriptText = "# Type a script or drag a script file from your workspace to insert its path.&#10;echo $PWD&#10;&#10;# Generate objc-friendly code&#10;./scripts/travis/generate_objc.sh&#10;&#10;">
```
Comment (codytwinton) [inline FacebookSwift.xcodeproj/xcshareddata/xcschemes/FacebookCore.xcscheme]: Prefer comments in the script file

```suggestion
               scriptText = "./scripts/travis/generate_objc.sh">
```

File: FacebookSwift.xcodeproj/xcshareddata/xcschemes/FacebookCore.xcscheme [hunk 4]
```diff
@@ -42,6 +70,15 @@
       debugDocumentVersioning = "YES"
       debugServiceExtension = "internal"
       allowLocationSimulation = "YES">
+      <PreActions>
+         <ExecutionAction
+            ActionType = "Xcode.IDEStandardExecutionActionsCore.ExecutionActionType.ShellScriptAction">
+            <ActionContent
+               title = "Run Script"
+               scriptText = "# Type a script or drag a script file from your workspace to insert its path.&#10;./scripts/travis/generate_objc.sh&#10;">
```
Comment (codytwinton) [inline FacebookSwift.xcodeproj/xcshareddata/xcschemes/FacebookCore.xcscheme]:
```suggestion
               scriptText = "./scripts/travis/generate_objc.sh">
```

File: IntegratingSDK/IntegratingSDK/Restaurant.m [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    FBRestaurant *restaurant = [[FBRestaurant alloc] initWithName: @"Mama Cass's Sandwich Shack"
+                                                         specials: @[@"Ham sammiches"]
+                                                      regularMenu: @[@"PB&Js"]
+                                                         employee: [[FBEmployee alloc] initWithClassification: payrollEntry]
+                                                          payroll: payrollEntry
+                                                     registerCode: @"123"
+                                                     bathroomCode: @"345"];
+
+    [restaurant ]
```
Comment (codytwinton) [inline IntegratingSDK/IntegratingSDK/Restaurant.m]: What is this?

File: IntegratingSDK/IntegratingSDK/Restaurant.m [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+-(FBRestaurant *)openRestaurant {
+    FBEmployeeClassification *payrollEntry = [[FBEmployeeClassification alloc] initWithCaseValue:[FBEmployeeClassificationManager new]];
+
+    FBRestaurant *restaurant = [[FBRestaurant alloc] initWithName: @"Mama Cass's Sandwich Shack"
+                                                         specials: @[@"Ham sammiches"]
+                                                      regularMenu: @[@"PB&Js"]
+                                                         employee: [[FBEmployee alloc] initWithClassification: payrollEntry]
+                                                          payroll: payrollEntry
+                                                     registerCode: @"123"
```
Comment (codytwinton) [inline IntegratingSDK/IntegratingSDK/Restaurant.m]: Should this not be an int or NSNumber?

File: Sources/Core/CustomStruct.swift [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  }
+  // Public Method that takes an enum
+
+  // Public Method that returns an enum
+
+  // Public Method that returns a bridged type
+}
+
+// Phantom type to mark types that may be bridged to objc
+protocol ObjCBridgeable {}
```
Comment (codytwinton) [inline Sources/Core/CustomStruct.swift]: Shouldn't this be in its own file? Something like:

```swift
// ObjCBridgeable.swift

// MARK: - Protocol

protocol ObjCBridgeable {}

// MARK: - Bridgeable

extension CustomStruct: ObjCBridgeable {}
extension Employee: ObjCBridgeable {}
extension EmployeeClassification: ObjCBridgeable {}
extension EmployeePayroll: ObjCBridgeable {}
extension Graph: ObjCBridgeable {}
extension Parser: ObjCBridgeable {}
extension StructWithGeneratedInitializer: ObjCBridgeable {}
```

File: Sources/Core/CustomStruct.swift [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+extension CustomStruct: ObjCBridgeable {}
```
Comment (codytwinton) [inline Sources/Core/CustomStruct.swift]: Shouldn't this be in its own file?

File: .gitattributes (hunk 1 above)
Comment (joesus) [reply .gitattributes]: I think that is correct yes.

File: .sourcery.yml (hunk 2 above)
Comment (joesus) [reply .sourcery.yml]: I guess it wouldn't hurt anything.

File: IntegratingSDK/IntegratingSDK/Restaurant.m (hunk 5 above)
Comment (joesus) [reply IntegratingSDK/IntegratingSDK/Restaurant.m]: Poor naming on my part. Fixing. Good catch!

File: IntegratingSDK/IntegratingSDK/Restaurant.m (hunk 6 above)
Comment (joesus) [reply IntegratingSDK/IntegratingSDK/Restaurant.m]: I see, I think there are two sample projects, I thought I had deleted one but I think it was just removed from the workspace. Apologies, I'll do some file cleanup.

File: Sources/Core/CustomStruct.swift (hunk 7 above)
Comment (joesus) [reply Sources/Core/CustomStruct.swift]: I like that idea. Making the change.

File: .gitattributes [hunk 9]
```diff
@@ -0,0 +1 @@
+Codegen/Generated/* linguist-generated=true
```
Comment (codytwinton) [inline .gitattributes]:
```suggestion
Codegen/Generated/* linguist-generated=true
*.stencil linguist-language=html+django
```

**Pull request merged**